### PR TITLE
Use FastHTTP client to improve HTTP performances 

### PR DESCRIPTION
**Pull request by kevinpollet**
### What does this PR do?

This pull request improves the HTTP performances between Traefik and the backends by using [FastHTTP](https://github.com/valyala/fasthttp) client primitives, for writing/reading HTTP requests as well as responses. The new HTTP proxy implementation using the FastHTTP proxy will only be used when making HTTP/1.1 requests to the backends.

This pull request also brings the following breaking changes:
- `passHostHeader` option is now a `ServersTransport` config option
- `responseForwarding.flushInterval` option has been removed as chunked responses are now handled automatically
- HTTP/2 is now disabled by default for communications between Traefik and the backends, unless the `enableHTTP2` option is set to `true` or the server URL uses the `h2c` scheme   
- To be consistent with other resources the `ServersTransport` configuration in the static configuration has been removed. It is now possible to configure a `default` ServersTransport as explained in the documentation.

### Benchmarks

Before:
```
Running 20s test @ http://traefik/bench
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.51ms    0.94ms  23.57ms   77.31%
    Req/Sec    17.44k     1.32k   51.46k    99.63%
  Latency Distribution
     50%    1.32ms
     75%    1.91ms
     90%    2.63ms
     99%    4.85ms
  1390186 requests in 20.10s, 135.23MB read
Requests/sec:  69166.15
Transfer/sec:      6.73MB
```

After:
```
Running 20s test @ http://traefik/bench
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.12ms  675.32us  17.63ms   78.60%
    Req/Sec    23.28k     1.97k   77.69k    99.63%
  Latency Distribution
     50%    1.00ms
     75%    1.41ms
     90%    1.89ms
     99%    3.52ms
  1855455 requests in 20.10s, 180.49MB read
Requests/sec:  92314.15
Transfer/sec:      8.98MB
```

### Motivation

To have better performances.

### More

- [X] Added/updated tests
- [X] Added/updated documentation

### Additional Notes

Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>

### CODE REVIEW

File: docs/content/migration/v2.md [hunk 1]
```diff
@@ -429,7 +429,7 @@ the legacy behavior related to the CommonName field can not be enabled at all an
 
 ### Errors middleware
 
-In `v2.5.4`, when the errors service is configured with the [`PassHostHeader`](../routing/services/index.md#pass-host-header) option to `true` (default),
+In `v2.5.4`, when the errors service is configured with the [`PassHostHeader`](https://doc.traefik.io/traefik/v2.5/routing/services/#pass-host-headerr) option to `true` (default),
```
Comment (tomMoulard) [inline docs/content/migration/v2.md:432]:
```suggestion
In `v2.5.4`, when the errors service is configured with the [`PassHostHeader`](https://doc.traefik.io/traefik/v2.5/routing/services/#pass-host-header) option to `true` (default),
```

File: docs/content/migration/v2-to-v3.md [hunk 2]
```diff
@@ -66,6 +66,24 @@ When using the KubernetesCRD provider, it is therefore necessary to update [RBAC
...
+
+In v3, the `PassHostHeader` config option has been moved to the [ServersTransport](../routing/services/index.md#passhostheader) resource.
+
+## ResponseForwarding.flushInterval
+
+In v3, the `ResponseForwarding.flushInterval` config option has been removed as chunked responses are now handled automatically.
+
+## HTTP/2
+
+In v3, HTTP/2 is disabled by default between Traefik and the backends unless it is enabled in the configured [ServersTransport](../routing/services/index.md#passhostheader)
```
Comment (rtribotte) [inline docs/content/migration/v2-to-v3.md:79]:
```suggestion
In v3, HTTP/2 is disabled by default between Traefik and the backends unless it is enabled in the configured [ServersTransport](../routing/services/index.md#enablehttp2) 
```

File: docs/content/migration/v2-to-v3.md [hunk 3]
```diff
@@ -66,6 +66,24 @@ When using the KubernetesCRD provider, it is therefore necessary to update [RBAC
...
+In v3, the `ResponseForwarding.flushInterval` config option has been removed as chunked responses are now handled automatically.
+
+## HTTP/2
+
+In v3, HTTP/2 is disabled by default between Traefik and the backends unless it is enabled in the configured [ServersTransport](../routing/services/index.md#passhostheader) 
+or the backend URL uses the `h2c` scheme.
+
+## ServersTransport
+
+In v3, the `default` [ServersTransport](../routing/services/index.md) should now be configured in the dynamic configuration.
```
Comment (rtribotte) [inline docs/content/migration/v2-to-v3.md:84]:
```suggestion
In v3, the `default` [ServersTransport](../routing/services/index.md#serverstransport_1) should now be configured in the dynamic configuration.
```

File: docs/content/migration/v2-to-v3.md [hunk 4]
```diff
@@ -66,6 +66,24 @@ When using the KubernetesCRD provider, it is therefore necessary to update [RBAC
...
+The ServersTransport resource now contains the [tls](../routing/services/index.md#tls) and [http](../routing/services/index.md#http) top level config options.
```
Comment (rtribotte) [inline docs/content/migration/v2-to-v3.md:85]:
```suggestion
The ServersTransport resource now contains a [tls](../routing/services/index.md#tls) top-level configuration option regrouping all the TLS-relative configuration options.  
```

**Comment by joeky888** [comment]
How about using [hertz](https://github.com/cloudwego/hertz) from ByteDance? It uses io_uring under the hood and supports http/2, performance is slightly better than fasthttp.

**Comment by mpl** [comment]
Hello @joeky888  ,

We haven't investigated/tried hertz because we're satisfied with what we have currently achieved with FastHTTP, and it would probably imply some additional significant effort to redo with hertz. So unless there's a really compelling argument to follow that lead, we're inclined to keep things as they are for now.
But thanks for the heads up nonetheless.

**Comment by joeky888** [comment]
Sure! We hope to see an ingress controller based on io_uring in the future. Hertz is still young, and [monoio](https://github.com/bytedance/monoio) is far from production-ready, both are developed by ByteDance.

**Comment by gchiesa** [comment]
is there any plan to merge this? I'm looking forward to this enhancement.

**Comment by nmengin** [comment]
I move the feature to the `next` milestone.
According to the [v3 migration plan](https://github.com/traefik/traefik/issues/10260#issuecomment-1867474446), with the maintainers, we do not want to bring such a breaking feature in v3.0.

We prefer to postpone this feature, make it non-breaking, and embed it in a future v3 minor version.

File: pkg/proxy/fasthttp/builder.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		dialTimeout = time.Duration(config.ForwardingTimeouts.DialTimeout)
+	}
+
+	proxyDialer := dialer.NewDialer(dialer.Config{
+		DialKeepAlive: 0,
+		DialTimeout:   dialTimeout,
+		HTTP:          true,
+		TLS:           targetURL.Scheme == "https",
+		ProxyURL:      proxyURL,
+	}, tlsConfig)
```
Comment (juliens) [inline pkg/proxy/fasthttp/builder.go:73]: Just a reminder to handle this issue #10403

**Comment by shkarface** [comment]
Now that v3 is officially released, @nmengin when do you think is a good time to look into this issue?

**Comment by nmengin** [comment]
Hello @shkarface,

Thank you for reaching out.
Indeed, we have planned to tackle this topic soon.

Note that this feature will be brought in a minor release, and it won't be enabled by default.

**Comment by kevinpollet** [comment]
Superseded by https://github.com/traefik/traefik/pull/11122

**Pull request closed without merging**